core/src/main/java/org/apache/iceberg/DistributedDataBatchScan.java

```diff
@@ -0,0 +1,309 @@
+
+      } catch (CompletionException e) {
+        dataFilesFuture.cancel(true /* may interrupt */);
+        deletesFuture.cancel(true /* may interrupt */);
+        throw new RuntimeException("Failed to plan files", e);
+
+      } finally {
+        workerPool.shutdown();
+        releaseResources();
+      }
+
+    } else {
+      return planFilesLocally(dataManifests, deleteManifests);
+    }
+  }
+
+  private CompletableFuture<CloseableIterable<DataFile>> newDataFilesFuture(
+      List<ManifestFile> manifests, boolean planRemotely, ExecutorService workerPool) {
+
+    return CompletableFuture.supplyAsync(
+        () -> {
+          if (planRemotely) {
+            scanMetrics().scannedDataManifests().increment(manifests.size());
+            List<DataFile> dataFiles = planDataRemotely(manifests);
+            int skippedDataFilesCount = totalFilesCount(manifests) - dataFiles.size();
+            scanMetrics().skippedDataFiles().increment(skippedDataFilesCount);
+            return CloseableIterable.withNoopClose(dataFiles);
+          } else {
+            ManifestGroup manifestGroup = newManifestGroup(manifests, ImmutableList.of());
+            return CloseableIterable.transform(manifestGroup.entries(), ManifestEntry::file);
+          }
+        },
+        workerPool);
+  }
+
+  private CompletableFuture<DeleteFileIndex> newDeletesFuture(
+      List<ManifestFile> manifests, boolean planRemotely, ExecutorService workerPool) {
+
+    return CompletableFuture.supplyAsync(
+        () -> {
+          DeleteFileIndex.Builder builder;
+
+          if (planRemotely) {
+            scanMetrics().scannedDeleteManifests().increment(manifests.size());
+            List<DeleteFile> deleteFiles = planDeletesRemotely(manifests);
+            int skippedDeleteFilesCount = totalFilesCount(manifests) - deleteFiles.size();
+            scanMetrics().skippedDeleteFiles().increment(skippedDeleteFilesCount);
+            builder = DeleteFileIndex.builderForFiles(io(), deleteFiles);
+          } else {
+            builder = DeleteFileIndex.builderFor(io(), manifests);
+          }
+
+          return builder
+              .specsById(table().specs())
+              .filterData(filter())
+              .caseSensitive(isCaseSensitive())
+              .planWith(planExecutor())
+              .scanMetrics(scanMetrics())
+              .build();
+        },
+        workerPool);
+  }
+
+  private CloseableIterable<ScanTask> planFilesLocally(
+      List<ManifestFile> dataManifests, List<ManifestFile> deleteManifests) {
+    ManifestGroup manifestGroup = newManifestGroup(dataManifests, deleteManifests);
+    return asScanTasks(manifestGroup.planFiles());
+  }
+
+  @Override
+  public CloseableIterable<ScanTaskGroup<ScanTask>> planTasks() {
+    return TableScanUtil.planTaskGroups(
+        planFiles(), targetSplitSize(), splitLookback(), splitOpenFileCost());
+  }
+
+  private CloseableIterable<ScanTask> createTasks(
```

Comment:
Minor: It isn't clear from the name that this is creating FileScanTask rather than a combined task. Normally, we'd use "files" vs "tasks" for that distinction. I think if this were instead `createFileTasks` it would be clear.

Reply:
Done.